### PR TITLE
Add multiset find APIs

### LINKED ISSUE #464: [FEATURE]: Add multiset host-bulk `find` and `find_async`
### Is your feature request related to a problem? Please describe.

 Add multiset host-bulk `find` and `find_async`

### Describe the solution you'd like

Two APIs to add:
```cuda
  /**
   * @brief For all keys in the range `[first, last)`, finds an element with key equivalent to the
   * query key.
   *
   * @note This function synchronizes the given stream. For asynchronous execution use `find_async`.
   * @note If the key `*(first + i)` has a matched `element` in the multiset, copies `element` to
   * `(output_begin + i)`. Else, copies the empty key sentinel.
   *
   * @tparam InputIt Device accessible input iterator
   * @tparam OutputIt Device accessible output iterator assignable from the set's `key_type`
   *
   * @param first Beginning of the sequence of keys
   * @param last End of the sequence of keys
   * @param output_begin Beginning of the sequence of elements retrieved for each key
   * @param stream Stream used for executing the kernels
   */
  template <typename InputIt, typename OutputIt>
  void find(InputIt first, InputIt last, OutputIt output_begin, cuda_stream_ref stream = {}) const;

  /**
   * @brief For all keys in the range `[first, last)`, asynchronously finds an element with key
   * equivalent to the query key.
   *
   * @note If the key `*(first + i)` has a matched `element` in the multiset, copies `element` to
   * `(output_begin + i)`. Else, copies the empty key sentinel.
   *
   * @tparam InputIt Device accessible input iterator
   * @tparam OutputIt Device accessible output iterator assignable from the set's `key_type`
   *
   * @param first Beginning of the sequence of keys
   * @param last End of the sequence of keys
   * @param output_begin Beginning of the sequence of elements retrieved for each key
   * @param stream Stream used for executing the kernels
   */
  template <typename InputIt, typename OutputIt>
  void find_async(InputIt first,
                  InputIt last,
                  OutputIt output_begin,
                  cuda_stream_ref stream = {}) const;
```

### Describe alternatives you've considered

_No response_

### Additional context

- This is the same as `static_set::find/find_async` but those two APIs are not directly accessible, need to find a way to expose those detail interfaces.
- Overloads taking custom hasher and comparator are not needed for now

### PR DESCRIPTION
**Pull request by PointKernel**
Closes #464 

This PR adds multiset host-bulk and device-singular find APIs

### CODE REVIEW

File: include/cuco/detail/static_multiset/static_multiset_ref.inl [hunk 1]
```diff
@@ -267,5 +267,68 @@ class operator_impl<
...
 
+template <typename Key,
+          cuda::thread_scope Scope,
+          typename KeyEqual,
+          typename ProbingScheme,
+          typename StorageRef,
+          typename... Operators>
+class operator_impl<
+  op::find_tag,
+  static_multiset_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>> {
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multiset/static_multiset_ref.inl:375]: The `find` operator should be the same for all of the data structures, right? Can we unify the implementation somehow to reduce code duplication?

File: include/cuco/detail/static_multiset/static_multiset_ref.inl [hunk 2]
```diff
@@ -267,5 +267,68 @@ class operator_impl<
...
+  using const_iterator = typename base_type::const_iterator;
+
+  static constexpr auto cg_size     = base_type::cg_size;
+  static constexpr auto window_size = base_type::window_size;
+
+ public:
+  /**
+   * @brief Finds an element in the multiset with key equivalent to the probe key.
+   *
+   * @note Returns a un-incrementable input iterator to the element whose key is equivalent to
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multiset/static_multiset_ref.inl:391]: Isn't a un-incrementable iterator the same as a reference?

File: tests/static_multiset/find_test.cu [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+#include <thrust/device_vector.h>
+#include <thrust/functional.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/sequence.h>
+#include <thrust/transform.h>
+
+#include <catch2/catch_template_test_macros.hpp>
+
+#include <cuda/functional>
```
Comment (sleeepyjack) [inline tests/static_multiset/find_test.cu]: Move to thrust include block

File: tests/static_multiset/find_test.cu [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+using size_type = int32_t;
+
+template <typename Set>
+__inline__ void test_unique_sequence(Set& set, size_type num_keys)
```
Comment (sleeepyjack) [inline tests/static_multiset/find_test.cu]:
```suggestion
void test_unique_sequence(Set& set, size_type num_keys)
```

Maybe superseeded by #481 .
If so, merging `dev` into this branch should remove the inline specifier.

File: include/cuco/detail/static_multiset/static_multiset_ref.inl (hunk 2 above)
Comment (PointKernel) [reply include/cuco/detail/static_multiset/static_multiset_ref.inl:391]: yes, what we return here is basically a constant iterator.

File: include/cuco/detail/static_multiset/static_multiset_ref.inl (hunk 1 above)
Comment (PointKernel) [reply include/cuco/detail/static_multiset/static_multiset_ref.inl:375]: very good point. Let me look into this.

File: tests/static_multiset/find_test.cu (hunk 3 above)
Comment (sleeepyjack) [reply tests/static_multiset/find_test.cu]: This one is still TBD

**Review by sleeepyjack — APPROVED** [review]
LGTM apart from the minor include reordering thing.

File: tests/static_multiset/find_test.cu (hunk 3 above)
Comment (sleeepyjack) [inline tests/static_multiset/find_test.cu]:
```suggestion
#include <cuda/functional>
#include <thrust/device_vector.h>
#include <thrust/functional.h>
#include <thrust/iterator/constant_iterator.h>
#include <thrust/sequence.h>
#include <thrust/transform.h>

#include <catch2/catch_template_test_macros.hpp>
```

File: include/cuco/detail/static_multiset/static_multiset.inl [hunk 5]
```diff
@@ -245,6 +246,42 @@ void static_multiset<Key, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Sto
...
+  InputIt first, InputIt last, OutputIt output_begin, cuda_stream_ref stream) const
+{
+  auto const num_keys = cuco::detail::distance(first, last);
+  if (num_keys == 0) { return; }
+
+  auto const grid_size = cuco::detail::grid_size(num_keys, cg_size);
+
+  static_set_ns::detail::find<cg_size, cuco::detail::default_block_size()>
+    <<<grid_size, cuco::detail::default_block_size(), 0, stream>>>(
+      first, num_keys, output_begin, ref(op::find));
```
Comment (PointKernel) [inline include/cuco/detail/static_multiset/static_multiset.inl:282]: this needs to be updated to use the OA `find`, I will open a separate PR for it.

Comment (sleeepyjack) [reply include/cuco/detail/static_multiset/static_multiset.inl:282]: Daang, I forgot about this part. Sorry, my bad

**Pull request merged**